### PR TITLE
[state_dict][10/N] Let set_state_dict returns IncompatibleKeys

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112414
* #112203
* #112167

load_state_dict returns IncompatibleKeys, so set should also return the same information for the users.

Differential Revision: [D50748157](https://our.internmc.facebook.com/intern/diff/D50748157/)

### CODE REVIEW

File: test/distributed/checkpoint/e2e/test_pipeline.py [hunk 1]
```diff
@@ -90,7 +90,7 @@ def load_with_fsdp(self, pipeline_dir: str) -> None:
             optim_state_dict=optim_state_dict,
         )
 
-    @skip_if_lt_x_gpu(4)
+    @skip_if_lt_x_gpu(8)
```
Comment (wz337) [inline test/distributed/checkpoint/e2e/test_pipeline.py:93]: Just curious, do multi-gpu CI tests have 8 gpus? Does 4 not enough for this test case?

**Review by wz337 — APPROVED** [review]
LGTM!

File: test/distributed/checkpoint/e2e/test_pipeline.py (hunk 1 above)
Comment (fegin) [reply test/distributed/checkpoint/e2e/test_pipeline.py:93]: oh, right, I was trying something else. Let me fix this.

**Pull request closed without merging**